.travis.yml

```diff
@@ -5,35 +5,47 @@ env:
 
+services:
+  - docker
+
 # todo: remove this once travis fixes their upstream issue (travis-ci/travis-ci#8836)
 sudo: required
-addons:
-  chrome: stable
 
-script:
-  # Bazel build & test everything. Can't use `bazel build ...` yet, because
-  # javaharness doesn't build on Travis (needs working Android SDK).
-  - >
-    bazel build --noshow_progress --noshow_loading_progress
-    //particles/...
-    //src/...
-    //src_kt/...
-  - >
-    bazel test --noshow_progress --noshow_loading_progress --test_output=errors
-    //particles/...
-    //src/...
-    //src_kt/...
-  - npm run test-with-start
+# Bazel build & test everything. Can't use `bazel build ...` yet, because
+# javaharness doesn't build on Travis (needs working Android SDK).
+jobs:
+  include:
+    - stage: build
+      script: tools/run_in_docker "bazel --bazelrc=tools/travis.bazelrc build //src_kt/..."
+    - stage: test
+      script: tools/run_in_docker "bazel --bazelrc=tools/travis.bazelrc test //src_kt/..."
+      name: "Bazel Test: //src_kt/..."
+    - script: tools/run_in_docker  "bazel --bazelrc=tools/travis.bazelrc test //src/..."
+      name: "Bazel Test: //src/..."
+    - script: tools/run_in_docker "bazel --bazelrc=tools/travis.bazelrc test //particles/..."
+      name: "Bazel Test: //particles/..."
+    - script: tools/run_in_docker "npm run test-no-web"
+      name: "Node Test: Unit"
+#     # TODO(alxr): Add back into CI once docker issue fixed
+#    - script: tools/run_in_docker "npm run test-shells"
+#      name: "Node Test: Shells"
 
+# Use travis to cache a Bazel disk-cache.
+# This connects to the docker container via a volume mount (`-v /home/travis/disk-cache:/disk-cache/`).
+# Bazel will store artifacts here due to the `--disk_cache=/disk-cache/` argument.
 cache:
   npm: true
+  directories:
+    - /home/travis/build/PolymerLabs/arcs/disk-cache/
 
-# Thinking about adding install steps? Why not modify the tools/setup script instead?
-# The list command dumps the top-level packages with their versions, in case we need that for debugging.
-install:
-  - ./tools/setup
-  - npm list --depth=0
+# Pulling container from https://hub.docker.com/repository/docker/alxrsngrtn/arcs-travis-build
+before_install:
+  - docker pull alxrsngrtn/arcs-travis-build:latest
+  - mkdir -p ./disk-cache/
+  - touch .bazelrc
 
+install:
+  - tools/run_in_docker "npm install"
```

Comment:
Is this needed? The Dockerfile already includes `RUN npm install`

Reply:
Yes this is required just in case the requirements change.

Reply:
I don't understand, what do you mean? I thought that if the package.kson file changed, docker would re-run the npm install command, since it comes after?

Reply:
I haven't automated the docker build step, so this is necessary if the versions get out of sync. Later, we can remove this when the docker image creation gets automated.